docs: C2 release creation detailed design for review (#362)
camaraproject/ReleaseManagement: documentation/SupportingDocuments/CAMARA-Release-Workflow-and-Metadata-Concept.md

The workflow achieves these objectives through:

> - Structured **metadata files (in YAML)** as authoritative source for release planning and status

Reviewer: ```suggestion
- Structured **metadata files (in YAML)** as authoritative source for release planning and status
- Dedicated **Release Issue** to trigger and track release attempts and provide information on release state progress (replacing the manual Wiki-based release tracker)
```

- Dedicated **snapshot branches per release attempt** with automated preparation and validation
- **CI enforcement** on PRs to `main` ensuring correctness before merge
- **Branch protection** restricting snapshot branch changes to automation only

> - **Release PRs** into snapshot branches for collaborative review of documentation

Reviewer: ```suggestion
-  Normal PRs used for collaborative review and updates of documentation on review branch
- **Release PR** merges review branch into snapshot branch
```

Author: Line 26 updated with: 

"**Release-review branch** for collaborative documentation review, merged into snapshot via Release PR"

The release-review branch allows direct commits for documentation updates - no need for an additional PR layer.

| `main_contacts`        | GitHub handles of code owners or maintainers (per API in `release-plan.yaml`). |
| `main` branch          | Development branch. All content is work-in-progress (`version: wip`). |
| Maintenance branch     | Long-lived branch for maintaining older release cycles (e.g., `maintenance-r3`). See Appendix for details. |
| Snapshot branch        | Automation-owned branch per release attempt (e.g., `release-snapshot/r4.1-abc1234`). Contains mechanical changes. |

> | Release-review branch          | Human-editable branch for reviewable content (e.g., `release-review/r4.1-abc1234`). Merged into snapshot branch. |

Reviewer: ```suggestion
| Release-review branch          | Human-editable branch for reviewable content (e.g., `release-review/r4.1-abc1234`). Contains mechanical changes refinable through PRs. Eventually merged into snapshot branch. |
```

Author: Not applied - release-review branch contains reviewable content (CHANGELOG, README). Mechanical changes go to snapshot branch.

  - `maintenance-release`: For maintenance/hotfix releases from maintenance branches
- CI validates that target API statuses match the declared repository target release type.

### 2. `release-metadata.yaml` (on snapshot branch)

> Generated automatically from the release plan and committed to the snapshot branch. Serves as the authoritative source of truth for the release attempt. Preserved in each release tag, providing complete release history.

Author: applied

- Keeps readiness transparent across reviewers, contributors, and release managers
- Allows optional enforcement that metadata updates (e.g., status changes) must be done in dedicated PRs separate from implementation changes

### Step 2: Automated Snapshot Creation

Upon triggering the release (via `/create-snapshot` slash command in a release issue - maintainers+ can trigger):

- Validation runs against current HEAD before any branches are created
- A snapshot branch is created (e.g., `release-snapshot/r4.1-abc1234`) with SHA-based naming
- A release-review branch is created from the snapshot branch (e.g., `release-review/r4.1-abc1234`)
- A script or GitHub Action:
  - Sets exact API versions using `target_api_version` + auto-calculated extension (e.g., `-rc.2` based on consecutive numbering across API lifecycle)
  - Enforces CAMARA versioning rules: info.version matches tag, server URLs follow v0.x or vx patterns
  - Writes `release-metadata.yaml` to the snapshot branch (source of truth for release parameters)
  - Commits mechanical changes (versions, URLs) to snapshot branch

>   - Commits reviewable content (CHANGELOG, README, checklists) to release-review branch

Reviewer: ```suggestion
  - Commits automated updates to reviewable content (CHANGELOG, README, checklists) to release-review branch
```

  - Opens a Release PR from release-review branch to snapshot branch

**Rationale:**
- Avoids fragile manual editing of version fields and metadata
- Creates verified, reviewable release state before tag
- Separates mechanical changes (protected) from reviewable content (editable)

> > **Detailed Design**: See [CAMARA-Release-Creation-Detailed-Design.md](CAMARA-Release-Creation-Detailed-Design.md) for comprehensive design including state model, command set, and bot UX contract.

Author: Changed to "... and bot interaction patterns."

> ### Step 3: Review via Release PRs

Reviewer: adjust the same in the rest of this paragraph. add that Release PR is used to merge the release branch into the snapshot branch after the review is done and changes agreed through PRs merges into the release-review branch.

   - Snapshot branch content is immutable after creation
   - Reviewable content (CHANGELOG) is on the release-review branch
   - If API corrections needed: discard snapshot, fix main, create new snapshot

#### Phase 2: Draft Creation and Publication (after PR merge)

After approval and Release PR merge to snapshot branch:

1. Automation creates draft GitHub Release (no tag yet)
2. Populates final metadata:
   - `release_date`: Current UTC timestamp
3. Human reviews draft release artifacts
4. **Human publishes** the release → creates git tag (e.g., `r4.1`)
5. CI builds and publishes artifacts
6. GitHub Release with artifacts is finalized:
   - Bundled OpenAPI specifications (later phase)
   - Generated documentation
   - Release metadata files

**Note:** Tag `rX.Y` is only created at final publication, not after PR merge. This provides a human checkpoint before publication.

**Rationale:**
- Provides traceable, repeatable state for each tagged release
- Two-phase approach ensures metadata reflects actual release state
- Human checkpoint prevents accidental publication
- Immutable snapshots simplify validation and prevent drift

> ### Step 5: Post-Release Actions

Author: 5a) eventually automated yes - see https://github.com/camaraproject/ReleaseManagement/issues/355
5c) we have the necessary data to create the tag at the right commit in the release-metadata.yaml, so can be done even years later. But we don't need the tags for the same reason at all, see https://github.com/camaraproject/ReleaseManagement/issues/366

- Reference for comparing API changes in next release
- Note: This is NOT a release tag, just a reference marker

#### 5d. Cleanup snapshot and release-review branches (optional):
- Snapshot branches (e.g., `release-snapshot/r4.1-abc1234`) are temporary scaffolding
- Release-review branches (e.g., `release-review/r4.1-abc1234`) may be kept for reference

> - The git tag preserves the release permanently

Reviewer: great we don't need these tags as they confused me :-)

- **Branch**: A movable pointer to commits, used for ongoing work
- **Tag**: A permanent marker on a specific commit, used to mark releases
- **Snapshot branch**: A temporary, automation-owned branch for preparing a specific release attempt
- **Release-review branch**: A temporary branch for reviewable content, merged into the snapshot branch
- **Maintenance branch**: A long-lived branch for maintaining older release cycles

> ### Snapshot Branch Lifecycle

Reviewer: do we need this snapshot branch life cycle, ? it seems so be the same as the release state (with a few different values) ? or was that the replacement in the detailed design ?
